src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -150,6 +152,82 @@ CorProfiler::Initialize(IUnknown* cor_profiler_info_unknown) {
   return S_OK;
 }
 
+HRESULT STDMETHODCALLTYPE CorProfiler::AssemblyLoadFinished(AssemblyID assembly_id,
+    HRESULT hr_status) {
+  if (FAILED(hr_status)) {
+    // if assembly failed to load, skip it entirely,
+    // otherwise we can crash the process if module is not valid
+    CorProfilerBase::AssemblyLoadFinished(assembly_id, hr_status);
+    return S_OK;
+  }
+
+  if (!is_attached_) {
+    return S_OK;
+  }
+
+  if (debug_logging_enabled) {
+    Debug("AssemblyLoadFinished: ", assembly_id, " ", hr_status);
+  }
+
+  // keep this lock until we are done using the module,
+  // to prevent it from unloading while in use
```

Comment:
Did you notice this happening?

Reply:
Nope, I just followed the convention of holding onto a lock while using the profiler like in the ModuleLoadFinished handler.

Reply:
Any deadlock chances?

Reply:
No. Only one of these callbacks is running at any point in time